File: lab3/gr8sub1/Павлючук_Юлия/src/thread_race.py
#!/usr/bin/env python3
import threading
import time
import sys
from threading import Lock

class Counter:
    def __init__(self):
        self.value = 0

class UnsynchronizedCounter(Counter):
    def increment(self):
        self.value += 1

class MutexCounter(Counter):
    def __init__(self):
        super().__init__()
        self.lock = Lock()
    
    def increment(self):
        with self.lock:
            self.value += 1

class AtomicCounter(Counter):
    def __init__(self):
        super().__init__()
        self.lock = Lock()  # Имитация атомарности
    
    def increment(self):
        with self.lock:
            self.value += 1

def worker_thread(counter, increments_per_thread):
    for _ in range(increments_per_thread):
        counter.increment()

def run_experiment(num_threads, total_increments, mode):
    if mode == "unsync":
        counter = UnsynchronizedCounter()
    elif mode == "mutex":
        counter = MutexCounter()
    elif mode == "atomic":
        counter = AtomicCounter()
    else:
        raise ValueError(f"Unknown mode: {mode}")
    
    increments_per_thread = total_increments // num_threads
    
    threads = []
    start_time = time.time()
    
    for i in range(num_threads):
        thread = threading.Thread(target=worker_thread, args=(counter, increments_per_thread))
        threads.append(thread)
        thread.start()
    
    for thread in threads:
        thread.join()
    
    end_time = time.time()
    execution_time_ms = (end_time - start_time) * 1000
    
    print(f"mode={mode} threads={num_threads} expected={total_increments} actual={counter.value} time_ms={execution_time_ms:.2f}")
    
    return execution_time_ms, counter.value

def main():
    if len(sys.argv) != 4:
        print("Usage: python thread_race.py <N> <M> <mode>")
        print("  N - number of threads")
        print("  M - total increments") 
        print("  mode - unsync, mutex, or atomic")
        sys.exit(1)
    
    num_threads = int(sys.argv[1])
    total_increments = int(sys.argv[2])
    mode = sys.argv[3]
    
    run_experiment(num_threads, total_increments, mode)

if __name__ == "__main__":
    main()

File: lab3/gr9sub2/БОГДЕВИЧ_АР/src/thread_race.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>
#include <string.h>
#include <time.h>
#include <stdatomic.h>

typedef struct {
    long long counter;
    pthread_mutex_t mutex;
    atomic_long atomic_counter;
    int num_threads;
    long long increments_per_thread;
    char mode[20];
} shared_data_t;

typedef struct {
    int thread_id;
    shared_data_t* shared_data;
} thread_data_t;

void* thread_func_unsync(void* arg) {
    thread_data_t* data = (thread_data_t*)arg;
    shared_data_t* shared = data->shared_data;
    
    for (long long i = 0; i < shared->increments_per_thread; i++) {
        shared->counter++;
    }
    
    return NULL;
}

void* thread_func_mutex(void* arg) {
    thread_data_t* data = (thread_data_t*)arg;
    shared_data_t* shared = data->shared_data;
    
    for (long long i = 0; i < shared->increments_per_thread; i++) {
        pthread_mutex_lock(&shared->mutex);
        shared->counter++;
        pthread_mutex_unlock(&shared->mutex);
    }
    
    return NULL;
}

void* thread_func_atomic(void* arg) {
    thread_data_t* data = (thread_data_t*)arg;
    shared_data_t* shared = data->shared_data;
    
    for (long long i = 0; i < shared->increments_per_thread; i++) {
        atomic_fetch_add_explicit(&shared->atomic_counter, 1, memory_order_relaxed);
    }
    
    return NULL;
}

int main(int argc, char* argv[]) {
    if (argc != 4) {
        printf("Usage: %s <num_threads> <increments_per_thread> <unsync|mutex|atomic>\n", argv[0]);
        return 1;
    }
    
    int num_threads = atoi(argv[1]);
    long long increments_per_thread = atoll(argv[2]);
    char* mode = argv[3];
    
    shared_data_t shared_data;
    shared_data.counter = 0;
    shared_data.num_threads = num_threads;
    shared_data.increments_per_thread = increments_per_thread;
    strcpy(shared_data.mode, mode);
    atomic_init(&shared_data.atomic_counter, 0);
    
    pthread_mutex_init(&shared_data.mutex, NULL);
    
    pthread_t* threads = malloc(num_threads * sizeof(pthread_t));
    thread_data_t* thread_data = malloc(num_threads * sizeof(thread_data_t));
    
    clock_t start_time = clock();
    
    // Create threads
    for (int i = 0; i < num_threads; i++) {
        thread_data[i].thread_id = i;
        thread_data[i].shared_data = &shared_data;
        
        if (strcmp(mode, "unsync") == 0) {
            pthread_create(&threads[i], NULL, thread_func_unsync, &thread_data[i]);
        } else if (strcmp(mode, "mutex") == 0) {
            pthread_create(&threads[i], NULL, thread_func_mutex, &thread_data[i]);
        } else if (strcmp(mode, "atomic") == 0) {
            pthread_create(&threads[i], NULL, thread_func_atomic, &thread_data[i]);
        } else {
            printf("Unknown mode: %s\n", mode);
            free(threads);
            free(thread_data);
            return 1;
        }
    }
    
    // Wait for all threads to complete
    for (int i = 0; i < num_threads; i++) {
        pthread_join(threads[i], NULL);
    }
    
    clock_t end_time = clock();
    double execution_time = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;
    
    // Print results
    long long expected = num_threads * increments_per_thread;
    long long actual = (strcmp(mode, "atomic") == 0) ? atomic_load(&shared_data.atomic_counter) : shared_data.counter;
    
    printf("Mode: %s\n", mode);
    printf("Threads: %d\n", num_threads);
    printf("Increments per thread: %lld\n", increments_per_thread);
    printf("Expected counter: %lld\n", expected);
    printf("Actual counter: %lld\n", actual);
    printf("Execution time: %.4f seconds\n", execution_time);
    printf("Correct: %s\n", (expected == actual) ? "YES" : "NO");
    
    // Cleanup
    free(threads);
    free(thread_data);
    pthread_mutex_destroy(&shared_data.mutex);
    
    return 0;
}
